clients/spark/core/src/main/hadoop2/scala/io/treeverse/clients/conditional/S3ClientBuilder.scala

```diff
@@ -28,16 +30,49 @@ object S3ClientBuilder extends io.treeverse.clients.S3ClientBuilder {
         )
       } else None
 
+    initializeClient(configuration, credentialsProvider, region, bucket)
+  }
+
+  private def initializeClient(
+      config: ClientConfiguration,
+      credentialsProvider: Option[AWSCredentialsProvider],
+      region: String,
+      bucket: String
+  ): AmazonS3 = {
     val builder = AmazonS3ClientBuilder
       .standard()
-      .withClientConfiguration(configuration)
+      .withClientConfiguration(config)
       .withRegion(region)
-
     val builderWithCredentials = credentialsProvider match {
       case Some(cp) => builder.withCredentials(cp)
       case None     => builder
     }
+    val client = builderWithCredentials.build
+
+    if (validateClientCorrectness(client, bucket)) {
+      return client
+    }
+    val bucketRegion = getAWSS3Region(client, bucket)
+    builder.withRegion(bucketRegion).build
+  }
+
+  private def validateClientCorrectness(client: AmazonS3, bucket: String): Boolean = {
+    try {
+      client.headBucket(new HeadBucketRequest(bucket))
+      true
+    } catch {
+      case _: Exception => false
+    }
+  }
 
-    builderWithCredentials.build
+  private def getAWSS3Region(client: AmazonS3, bucket: String): String = {
+    val bucketRegion = client.getBucketLocation(bucket)
+    val region = Region.fromValue(bucketRegion)
+    // The comparison `region.equals(Region.US_Standard))` is required due to AWS's backward compatibility:
+    // https://github.com/aws/aws-sdk-java/issues/1470.
+    // "us-east-1" was previously called "US Standard". This resulted in a return value of "US" when
+    // calling `client.getBucketLocation(bucket)`.
```

Comment:
Thanks, very helpful explanation!

[Best Java SDK is Amazon AWS v1 Java SDK.
![Kim "best Korea" meme](https://user-images.githubusercontent.com/7802932/201023933-95faa668-70a1-46c2-8806-1f084023554c.png)](https://imgflip.com/i/707w5o)